### PR TITLE
Emitting tsserverlibrary as an external module.

### LINKED ISSUE #13236: No straightforward/stable way to consume lib/tsserverlibrary.js
**TypeScript Version:**  2.\*.\*-\*

**Code**
This is in a project that needs to use tsserverlibrary.js:

test.ts:
```ts
import * as ts from "typescript/lib/tsserverlibrary";
```
Because tsserverlibrary.d.ts is not a module:
```ts
/// <reference ...
declare module "typescript/lib/tsserverlibrary" {
    export = ts;
}
```

This still fails due to the following line in tsserverlibrary.d.ts, which in an installed package, resolves to the nonexistent file node_modules/src/types.d.ts:
```ts
/// <reference path="../../src/server/types.d.ts" />
```

With a custom copy of tsserverlibrary.d.ts, `const ts = require(...` is a potential option, but options for consuming `ts` as the declared namespace instead of `any` are clumsy.

**Expected behavior:**
Consumer of the TypeScript Node package can import tsserverlibrary.js without issue.

### PR DESCRIPTION
**Pull request by joelday**
Fixes #13236

Verified that compiled output of .js files is identical to master. "tsserverlibrary.d.ts" now exports `ts` and no longer has a triple slash reference to a d.ts file. Thanks!

### CODE REVIEW

**Comment by joelday** [comment]
🛑 Didn't catch that the d.ts isn't being generated properly. Fixing now.

File: src/server/tsconfig.library.json [hunk 1]
```diff
@@ -1,29 +1,33 @@
...
-        "utilities.ts",
-        "scriptVersionCache.ts",
-        "scriptInfo.ts",
-        "lshost.ts",
-        "typingsCache.ts",
-        "project.ts",
         "editorServices.ts",
+        "lsHost.ts",
+        "project.ts",
         "protocol.d.ts",
```
Comment (mhegazy) [inline src/server/tsconfig.library.json]: `protocol.ts`

File: src/server/tsconfig.library.json [hunk 2]
```diff
@@ -1,29 +1,33 @@
...
+        "removeComments": false,
         "preserveConstEnums": true,
+        "pretty": true,
         "outFile": "../../built/local/tsserverlibrary.js",
         "sourceMap": true,
-        "stripInternal": true,
-        "declaration": true,
-        "types": [],
+        "types": [
+            "node"
```
Comment (mhegazy) [inline src/server/tsconfig.library.json]: do not think you need `node` here. the library should not have any dependency on the environment, if it does, this is a bug that needs to be fixed.

File: src/server/tsconfig.library.json [hunk 3]
```diff
@@ -1,29 +1,33 @@
...
     "compilerOptions": {
         "noImplicitAny": true,
-        "removeComments": true,
+        "noImplicitThis": true,
+        "removeComments": false,
         "preserveConstEnums": true,
+        "pretty": true,
         "outFile": "../../built/local/tsserverlibrary.js",
         "sourceMap": true,
-        "stripInternal": true,
```
Comment (mhegazy) [inline src/server/tsconfig.library.json]: why is this removed?

File: src/server/tsconfig.library.json [hunk 4]
```diff
@@ -1,29 +1,33 @@
 {
     "compilerOptions": {
         "noImplicitAny": true,
-        "removeComments": true,
+        "noImplicitThis": true,
+        "removeComments": false,
```
Comment (mhegazy) [inline src/server/tsconfig.library.json]: remove. the default is `false`.

File: Jakefile.js [hunk 5]
```diff
@@ -727,7 +719,15 @@ compileFile(
     [builtLocalDirectory, copyright, builtLocalCompiler].concat(languageServiceLibrarySources).concat(libraryTargets),
     /*prefixes*/[copyright],
     /*useBuiltCompiler*/ true,
-    { noOutFile: false, generateDeclarations: true });
+    { noOutFile: false, generateDeclarations: true },
+            /*callback*/ function () {
```
Comment (mhegazy) [inline Jakefile.js]: nit, extra spaces

File: Jakefile.js [hunk 6]
```diff
@@ -727,7 +719,15 @@ compileFile(
...
     /*prefixes*/[copyright],
     /*useBuiltCompiler*/ true,
-    { noOutFile: false, generateDeclarations: true });
+    { noOutFile: false, generateDeclarations: true },
+            /*callback*/ function () {
+
+        prependFile(copyright, tsserverLibraryDefinitionFile);
+
+        // Appending 'export = ts;' at the end of the server library file to turn it into an external module
+        var tsserverLibraryDefinitionFileContents = fs.readFileSync(tsserverLibraryDefinitionFile).toString() + "\r\nexport = ts;";
```
Comment (mhegazy) [inline Jakefile.js]: this will beak users who have been using this file as global. we should add an additional line `export as namespace ts;`

**Comment by joelday** [comment]
@mhegazy Made changes. :) Let me know if there's anything else I should do/change.

**Comment by mhegazy** [comment]
thanks. merged manually.
